travis: split build into matrix of 3

matthewbauer opened the pull request:
###### Motivation for this change

Once "language: nix" is available, this can be used for builds on Travis CI.
###### Things done
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

This adds a 3rd matrix to be built by Travis. The new matrix "checks"
the NixPkgs evaluation so the other 2 can save their resources for
building. Hopefully, this will lead to less "out of space" errors that
seem to be happening with Travis.

matthewbauer committed:
travis: split build into matrix of 3

This adds a 3rd matrix to be built by Travis. The new matrix "checks"
the NixPkgs evaluation so the other 2 can save their resources for
building. Hopefully, this will lead to less "out of space" errors that
seem to be happening with Travis. Also adds folding.

wmertens commented on maintainers/scripts/travis-nox-review-pr.sh:
```diff
@@ -1,58 +1,74 @@
 #! /usr/bin/env bash
-set -e
+set -ev
```
Why add -v? Things are verbose enough imho, no need to add bash in the mix…

matthewbauer replied:
All of the Travis examples have some it: https://docs.travis-ci.com/user/customizing-the-build/#Implementing-Complex-Build-Steps
I was hoping that Travis would minimize it? I'll remove it for now because it's not really related to the PR though.

wmertens commented:
[![LGTM](http://lgtm.in/p/MsGtv2YLn)](http://lgtm.in/i/MsGtv2YLn)

matthewbauer commented:
I'm hoping Travis uses the new .travis.yml, but I'm not sure if that's how it work.

wmertens commented:
I'm going to give @domenkozar a chance to look at it, we can merge this tomorrow and see if it still works then :)

wmertens commented:
What surprises me is that in https://travis-ci.org/NixOS/nixpkgs/builds/140654568 the Darwin build worked fine, but in the subsequent ones it always fails, due to `payload['title']` apparently not having that key. Any idea what could be going on?

@madjar can you review this too?

matthewbauer commented:
I can't figure out why Nox keeps getting built… AFAICT it should be
available in the binary caches.

On Tue, Jun 28, 2016 at 1:22 AM Wout Mertens <notifications@github.com>
wrote:

> What surprises me is that in
> https://travis-ci.org/NixOS/nixpkgs/builds/140654568 the Darwin build
> worked fine, but in the subsequent ones it always fails, due to
> payload['title'] apparently not having that key. Any idea what could be
> going on?
>
> @madjar <https://github.com/madjar> can you review this too?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/16563#issuecomment-228959116>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AABKXApfIQHeU6UIYi4QlsV4xfol1NV_ks5qQL2ggaJpZM4I_h7l>
> .
>

wmertens commented:
maybe nix-shell uses the cloned nixpkgs somehow? Maybe it tries the current dir first?

matthewbauer commented:
It looks like it's related to madjar/nox#26.

domenkozar commented:
It's using `nox-0.0.2` while `nox-0.0.3` fixes the issue

matthewbauer commented:
nox-0.0.3 shold be in nixpkgs-unstable, right?

matthewbauer commented:
Okay nixpkgs-unstable is still at 0.0.2. https://github.com/NixOS/nixpkgs-channels/blob/nixpkgs-unstable/pkgs/tools/package-management/nox/default.nix

matthewbauer committed:
travis: only fetch top commit

This will make travis only fetch the top level commit instead of
fetching the top 50 (default). Nox should be able fetch other commits on
an as-needed basis.

madjar commented:
This PR is great!

The transient failure we are talking about is not really fixed in nox-0.0.3, and I explain in https://github.com/madjar/nox/issues/29#issuecomment-229607171 that it won't be easy to fix.

matthewbauer commented:
So I'm going to put this on hold until we can get the latest from nox pulled into master. I'm willing to sacrifice a personal access token if that's the only way to prevent the error. It might be safer to just create a dummy GitHub user and just use that. I can't imagine there's much you can do with a scopeless personal access token though.

matthewbauer committed:
travis: use community supported language

Switch to the "community supported" Nix version of Travis.

domenkozar commented:
I think this makes sense. It's easier to see what check actually failed without looking at logs.

matthewbauer commented:
Yeah, if we still get as many errors as we do now in Linux and OSX builds we could make the "nox-review pr" part optional because there's about a 50% chance the Travis build will just run out of space.

matthewbauer commented:
@madjar Could you bump the version of nox in nixpkgs in the next week or so so that I can put all of the Travis changes in one pull request?

madjar commented:
@matthewbauer Here you go: https://github.com/NixOS/nixpkgs/pull/16700. I'll let you merge it :)

matthewbauer committed:
travis: support nox 0.0.4 arguments

This adds the "slug" arguments and also the "token" argument. The slug
argument provides the "owner_name/repo_name" format base repo to use for
the pull request. The token argument provides the GitHub presonal access
token to use for the requests to the GitHub API.

matthewbauer committed:
travis: set GITHUB_TOKEN variable in .travis.yml

This should allow for more builds per hour and less errors from nox-review.

matthewbauer commented:
Ok, so once Travis gets "Nix" included as a language this should be okay to merge!

matthewbauer committed:
travis: build manuals to verify that they work

fixes #15109

matthewbauer committed:
travis: run nixpkgs-lint in build script

This will run nixpkgs-lint for each travis build. We'll disable
"nixpkgs-lint" for now, it was a little too verbose:

https://travis-ci.org/NixOS/nixpkgs/jobs/144290920

matthewbauer commented:
It looks like it's now available in production on Travis so this should be okay to merge!

https://travis-ci.org/matthewbauer/nixpkgs/jobs/144277096

garbas commented:
@matthewbauer shouldn't we encrypt that github token?

garbas commented:
@matthewbauer also i'm seeing we're hitting quite few times Github API limit. Does each Github API token get its own limit or is this tied to the Github account?

matthewbauer commented:
So, if we encrypt it, the variable will only be available to "trusted" builds. That excludes PR from a fork of nixpkgs which is what 99% of PRs to nixpkgs are. I decided to just make an account @nixtravis and create that personal token without any of the scopes so the damage done from it being public should be close to 0.

Without the API token, you get 60 requests/hour for each IP address. I think because Travis shares IP addresses it's hitting it much faster. Each GitHub token should get 6000 requests/hour so I think we should be okay sharing one for all Travis builds. We only need those when Nox builds pull requests so it shouldn't be an issue.

matthewbauer commented:
So everything seems to be working on some of the latest PRs:

https://travis-ci.org/NixOS/nixpkgs/builds/145050229

@domenkozar One thing I'm confused about is why Nix is trying to build some of these drvs on OS X instead of just downloading them:

```
these derivations will be built:
  /nix/store/10hbs719c0hl6lc7mgfli74cjscdgbs8-python3.5-characteristic-14.1.0.drv
  /nix/store/b9knv9dh1c98jiqg2l5cz1qw133a1vs0-git-2.9.1.drv
  /nix/store/y5lpr6grqrip3s6q1sz3zpmag3a32jn7-python3.5-dogpile.core-0.4.1.drv
  /nix/store/krzdmwdl25abmvc68nrda089ckgwlbnw-python3.5-dogpile.cache-0.5.4.drv
  /nix/store/gn5abmpxv0rlg19d9v60hiljh0kz1vz4-nox-0.0.4.drv
```

Shouldn't these be available in the binary cache?

matthewbauer commented:
Okay here's another one:

```
these derivations will be built:
  /nix/store/xi2g7593l27rlqsyq6b6jzzz9gb7kxn4-fontconfig-2.11.1.drv
  /nix/store/zcgq15si83ff1wy0xqcpyg1rc2sb5w7k-libXft-2.3.2.drv
  /nix/store/r6wv03mk53wgcvg6p33bnxl086l0sfiy-tk-8.6.4.drv
  /nix/store/2gfi1mp1cirwx1l04asrlqp9bl1a4453-python3-3.5.2.drv
  /nix/store/h51sh3iic8r6vl22yk52r4zg1q318q8l-hook.drv
  /nix/store/lc1nr4s3gw5nmf516miiwyrq3790v8cf-python-3.5.2-bootstrapped-pip-8.1.2.drv
  /nix/store/rz8javidgjay6dq6qkl4kl1nvn9bhwrl-python3.5m-setuptools-19.4.drv
  /nix/store/xx8qlcbn18fa0bshl6ycd1flfx787i7a-python3.5-py-1.4.31.drv
  /nix/store/jibbfz1ibxb9fbibqmnkqan757bnzj8c-python3.5-pytest-2.9.2.drv
  /nix/store/1k8x05ff2fb8i3hizihnp6avsjqpqjrm-python3.5-requests-2.9.1.drv
  /nix/store/3vswjlsqjwjz0mm200bfhri4g7qyr4mb-python3.5-characteristic-14.1.0.drv
  /nix/store/sx476is5v3kvn22r0pvskypk4n8apkx7-python3.5-dogpile.core-0.4.1.drv
  /nix/store/6kd4f1i6wml7bjzhyl6pvxdaqc1a8sn3-python3.5-dogpile.cache-0.5.4.drv
  /nix/store/8i0m3nmyxwhjyy7q7xlayxzhj1r65far-python3.5-pbr-1.8.1.drv
  /nix/store/9585mrb9a9s2b2qh0mpdc8ps5m993qd3-python3.5-click-6.6.drv
  /nix/store/b9knv9dh1c98jiqg2l5cz1qw133a1vs0-git-2.9.1.drv
  /nix/store/k23743dvs5hmsnixksrxhp7r8z5lly04-nox-0.0.4.drv
```

I don't know how the binary stores works to really understand this properly.

matthewbauer commented:
It looks like the issue is from using Nox and Git from master instead of from the channels.
